### PR TITLE
Switch to Floskell

### LINKED ISSUE #5: Switch to Floskell
## Description

We want to use [Floskell](https://hackage.haskell.org/package/floskell) instead of Brittany for formatting.

## Motivation

Currently we are using [Brittany](https://hackage.haskell.org/package/brittany) for formatting Haskell source code.
Brittany is one of the few Haskell formatters that breaks up long lines and whose output is independent of the formatting chosen by the user. Unfortunately, Brittany supports only very few language constructs (mainly function declarations). Data type declarations are not formatted and imports are not sorted. Furthermore there is a bug on Windows that causes carriage return characters in comments to be duplicated. Thus, we've looked for a different formatter.

Of those formatters we've tested, [Floskell](https://hackage.haskell.org/package/floskell) was the only formatter that produced satisfying results. Like Brittany, Floskell breaks up long lines. Since Floskell takes operator precedence into account, it acually breaks up lines better than Brittany. Additionally, Floskell formats data type declarations, sorts imports and has many configuration options.

## Tasks

The following tasks should be done in the `guidelines` repository.

 - [x] Add a template configuration file for Floskell that matches our coding style.
 - [x] Replace sections about Brittany in the `CONTRIBUTING.md` file by sections about Floskell.
 - [x] Update the Haskell Styleguide in `CONTRIBUTING.md`. Aspects covered by Floskell don't have to be mentioned anymore.
 - [x] Mention the exact Floskell version used by the CI pipeline in `CONTRIBUTING.md`.

Open an issue or pull request on all other repositories with the following tasks.

 - [x] Add a Floskell confiuguration file based on the template in the [`guidelines`](https://github.com/FreeProving/guidelines) repository. Adapt the configuration file based on the repository's needs (e.g., language extensions and prefixes for grouping of imports). Do not change formatting options. Code style should be consistent across all repositories. 
 - [x] Update the `./tool/check-formatting.sh` and `./tool/format-code.sh` (and in case of the Free Compiler `./tool/full-test.sh`) scripts to use Floskell. Preferably copy an updated version from another repository.
 - [x] Update the configuration of the CI pipeline. Preferably copy updated CI jobs from another repository.
 - [x] Format code using Floskell.
 - [x] Check the code for errors or odd formatting introduced by Floskell. Adapt the configuration file if necessary.

Additionally, we will have to publish a Docker image for the CI pipeline that has the right version of Floskell preinstalled. For Brittany, I've created [`just95/brittany`](https://hub.docker.com/r/just95/brittany). The Dockerfile for Floskell should require only very little changes.

### PR DESCRIPTION
**Pull request by just95**
### Issue

Closes #5.

### Description of the Change

 - A template for the Floskell configuration file has been added.
 - The constributing guidelines have been updated from Brittany to Floskell.

### Additional Notes

Since Floskell uses JSON as a configuration file format, but JSON files cannot contain comments, we are using [TOML](https://toml.io/en/v0.4.0) and provide TOML to JSON conversion utilities. In this case we allow the generated `floskell.json` file to be committed to ease the transition to Floskell.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -62,12 +62,31 @@ Slight adaptations will be necessary on a per repository basis as not all jobs o
...
 
+## Floskell
+
+We are using Floskell to format Haskell source code.
+How Floskell formats the source code can be configured with a `./floskell.json` file.
+We are providing a [template][guidelines/floskell.toml] for the configuration file in this repository.
+The template uses [TOML][toml/v0.4.0] instead of JSON in order to add comments to the individual configuration options.
+
+The template should NOT be copied to the other repositories.
+Instead every repository should provide a own `./floskell.toml` file that contains repository-specific options.
```
Comment (nbun) [inline README.md]:
```suggestion
Instead every repository should provide its own `./floskell.toml` file that contains repository-specific options.
```

File: floskell.toml [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+language = "Haskell2010"
+
+# Per-Repository Configuration
+# ----------------------------
+#
+# Some configuration options are repository-specific. Each repository should
+# contain a `floskell.toml` file that contains ONLY the repository-specific
+# options. When generating the `floskell.json` file the repository-specific
+# file should be merged with this template file. Merging and conversion to JSON
+# can be performed either automatically with the following command, for example.
```
Comment (nbun) [inline floskell.toml]:
```suggestion
# can be performed automatically with the following command, for example.
```

File: floskell.toml [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+#     docker run -v $PWD:/pwd just95/toml-to-json \
+#       /pwd/guidelines/floskell.toml             \
+#       /pwd/other-repo/floskell.toml
+#
+# The repository should contain a `./tool/update-floskell-config.sh` script
+# for this task.
+#
+# The following top-level key/value pairs are required.
+#
+#     * `extensions` (`Array` of `String`) the extension that are enabled for
```
Comment (nbun) [inline floskell.toml]:
```suggestion
#     * `extensions` (`Array` of `String`) the extensions that are enabled for
```

File: floskell.toml [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+#       file).
+#     * `fixities` (`Array` of `String`) a list of fixity declarations of the
+#       form `infix[l|r] <PRECEDENCE> <OPERATOR>`. Floskell formats individual
+#       modules and does not inspect imported modules. If a module imports an
+#       infix operator from another module, Floskell needs to know the
+#       associtivity and precedence of the operator to parse the module
+#       correctly. Thus, the fixities of all imported infix operators have
+#       to be listed here. Unfortunatly, this means that all infix operator
+#       names must be unique. Avoid the definition of custom infix operators
+#       as well as the use third-party infix operators to prevent name clashes.
```
Comment (nbun) [inline floskell.toml]:
```suggestion
#       as well as the use of third-party infix operators to prevent name clashes.
```

File: floskell.toml [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+#
+# Additionally, an entry of the `formatting.options.sort-imports` array of
+# tables us needed. See the comment at the bottom of this file for details.
```
Comment (nbun) [inline floskell.toml]:
```suggestion
# tables is needed. See the comment at the bottom of this file for details.
```

File: floskell.toml [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+    # The equal sign of multi-line function declarations appears on the same
+    # line if it fits.
+    declaration      = "flex"
+
+    # If all constructors of a data type declaration fit into a single line,
+    # no line breaks are inserted. Otherwise each constructor is on its own
+    # line.
+    con-decls        = "try-oneline"
+
+    # If a `if`- or `let`-expression does not fit in a single line, the `in`,
```
Comment (nbun) [inline floskell.toml]:
```suggestion
    # If a `if`- or `let`-expression does not fit in a single line, the `if`,
```

File: floskell.toml [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+    # `then` and `else` expressions are broken into their own lines.
+    # For example, `then` and `else` don't share a line unless the entire
+    # `if` expression fits on the line.
```
Comment (nbun) [inline floskell.toml]:
```suggestion
    # `if` expression fits in one line.
```

File: floskell.toml [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+    # The body of class and instance declarations is indented by two spaces.
+    class            = "indent-by 2"
+
+    # The first alternative of a `case`-expression or the first statement of a
+    # `do`-block starts on a new line and is indented by two spaces.
+    case = "indent-by 2"
+    do   = "indent-by 2"
+
+    # If an export or import specifier list does not fit into the same line as
+    # the module header or import declaration they are broken into the next
```
Comment (nbun) [inline floskell.toml]:
```suggestion
    # the module header or import declaration, they are broken into the next
```

File: floskell.toml [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+    deriving    = 1
+    where       = 1
+    where-binds = "indent-by 2"
+
+    # Floskell does not break `::` into the next line if a type signature does
+    # not fit into a single line. This option aligns the `->` of the subsequent
+    # arguments with the `::` of the first argument.
+    typesig = "align"
+
+    # If an application expression does not fit into or a line break has to be
```
Comment (nbun) [inline floskell.toml]:
```suggestion
    # If an application expression does not fit into a line or a line break has to be
```

File: floskell.toml [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+    # inserted anywhere else, the new line is indented by two spaces.
+    app              = "indent-by 2"
+    onside           = 2
+
+  # The following rules control tabstop alignment and alignment limits.
+  [formatting.align]
+    # The following rules enable tabstop alignment everywhere except for
+    # import specifier lists. In consequence the import specifier list of
+    # `Foo` is not aligned with the the list of `Bar.Baz` in the example
+    # below. The import specifier list always follows the moduke name if
```
Comment (nbun) [inline floskell.toml]:
```suggestion
    # below. The import specifier list always follows the module name if
```

File: floskell.toml [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+    class         = true
+    import-module = true
+    import-spec   = false
+    let-binds     = true
+    matches       = true
+    record-fields = true
+    where         = true
+
+    # The following option controls the limits of tabstop alignment.
+    # According to Floskells documentation the numbers are to be interpreted
```
Comment (nbun) [inline floskell.toml]:
```suggestion
    # According to Floskell's documentation the numbers are to be interpreted
```

File: floskell.toml [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+    # as follows.
+    #
+    # > the first integer is a number of characters (absolute limit) and the
+    # > second value is a percentage (relvative limit).
```
Comment (nbun) [inline floskell.toml]:
```suggestion
    # > second value is a percentage (relative limit).
```

File: floskell.toml [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+    # > larger than absolute limit and is larger than relative limit of the
+    # > maximum field length (tabstop column minus indentation).
+    #
+    # We are using the default configuration, i.e. an absolute limit of 10
+    # characters and a relative limit of 25%. We may want to tweak these
+    # values in the future, though.
+    limits = [10, 25]
+
+  # The following rules control where to insert whitespace in parentheses,
+  # backets and braces.
```
Comment (nbun) [inline floskell.toml]:
```suggestion
  # brackets and braces.
```

File: floskell.toml [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+      linebreaks      = "before"
+
+    # In lambda expressions and `case` expression alternatives, the arrow is
+    # not broken into the next line.
+    [formatting.op."-> in expression"]
+      force-linebreak = false
+      spaces          = "both"
+      linebreaks      = "after"
+
+  # The following options.
```
Comment (nbun) [inline floskell.toml]:
```suggestion
  # The formatting options.
```

**Review by nbun — CHANGES_REQUESTED** [review]

**Review by nbun — APPROVED** [review]

File: floskell.toml (hunk 6 above)
Comment (just95) [reply floskell.toml]: I actually meant the `in` keyword of `let` expressions. In our configuration, Floskell tries to keep the `if` and `let` keywords on the previous line if possible. Maybe the following could clear up the confusion.

```haskell
# If a `let`- or `if`-expression does not fit in a single line, the `in`,
# `then` and `else` keywords are broken into their own lines.
```

File: floskell.toml (hunk 14 above)
Comment (just95) [reply floskell.toml]: Oops, I think I forgot to complete the sentence :joy:

**Review by stu204146-kiel — APPROVED** [review]

**Pull request merged**